### PR TITLE
More tests for model, component and parameter

### PR DESCRIPTION
**Pull request by to266**
Things that can already be reviewed:
 - [x] New parameter tests
 - [x] New component tests
 - [x] New model tests

### CODE REVIEW

File: hyperspy/component.py [hunk 1]
```diff
@@ -910,20 +907,18 @@ def _component2plot(self, axes_manager, out_of_range2nans=True):
         if axes_manager is not self.model.axes_manager:
             old_axes_manager = self.model.axes_manager
             self.model.axes_manager = axes_manager
-            self.charge()
```
Comment (francisco-dlp) [inline hyperspy/component.py]: This was actually a bug: `charge` was renamed to `fetch_stored_values`, but not here. Removing it causes another issue. I would suggest to simply replace `charge` with `fetch_stored_values`.

File: hyperspy/component.py [hunk 2]
```diff
@@ -910,20 +907,18 @@ def _component2plot(self, axes_manager, out_of_range2nans=True):
...
         s = self.__call__()
         if not self.active:
             s.fill(np.nan)
         if self.model.spectrum.metadata.Signal.binned is True:
             s *= self.model.spectrum.axes_manager.signal_axes[0].scale
-        if old_axes_manager is not None:
-            self.model.axes_manager = old_axes_manager
-            self.charge()
```
Comment (francisco-dlp) [inline hyperspy/component.py]: Same as above.

File: hyperspy/model.py [hunk 3]
```diff
@@ -456,7 +451,7 @@ def remove(self, thing, touch=True):
             idx = self.index(thing)
             self.spectrum._plot.signal_plot.ax_lines.remove(
                 self.spectrum._plot.signal_plot.ax_lines[2 + idx])
-        list.remove(self, thing)
+        list.__delitem__(self, self.index(thing))
```
Comment (francisco-dlp) [inline hyperspy/model.py]: why is this better?

Comment (to266) [inline hyperspy/model.py]: Hm... don't remember why I changed it. Probably not better

File: hyperspy/component.py (hunk 1 above)
Comment (to266) [inline hyperspy/component.py]: I guess it's one of the plotting bugs? Now I can see what might be a problem, I'll add the lines shortly

**Comment by to266** [comment]
All done

File: hyperspy/model.py [hunk 4]
```diff
@@ -960,60 +946,20 @@ def reset_the_signal_range(self):
...
-                                counter:counter +
-                                component._nfree_param],
-                            axis)
-                    counter += component._nfree_param
-            to_return = sum
-
-        if self.spectrum.metadata.Signal.binned is True:
-            to_return *= self.spectrum.axes_manager[-1].scale
+        old_p0 = self.p0
+        self._set_p0()
```
Comment (francisco-dlp) [inline hyperspy/model.py]: Isn't setting p0 unnecessary?

File: hyperspy/model.py [hunk 5]
```diff
@@ -960,60 +946,20 @@ def reset_the_signal_range(self):
...
+        old_values = self.p0
```
Comment (francisco-dlp) [inline hyperspy/model.py]: Backing up p0 should also be unnecessary.

File: hyperspy/model.py [hunk 6]
```diff
@@ -960,60 +946,20 @@ def reset_the_signal_range(self):
...
+        self.p0 = param
+        self._fetch_values_from_p0()
+        to_return = self.__call__(non_convolved=False, onlyactive=True)
+        self.p0 = old_values
+        self._fetch_values_from_p0()
```
Comment (francisco-dlp) [inline hyperspy/model.py]: If the answer to the above comment is yes, then this line can also go.

Comment (francisco-dlp) [inline hyperspy/model.py]: Now we can get rid of `Component.__tempcall__` :)

Comment (to266) [inline hyperspy/model.py]: All this play was only to, in the end of the function call, have the same `self.p0` and all `values`. If it's actually not necessary (i'm not sure if we break anything), then yes, only fetching the `param` values should be enough.

I'll try to make it so and see if anything breaks

Comment (to266) [inline hyperspy/model.py]: And it wasn't :) done

File: hyperspy/component.py (hunk 2 above)
Comment (francisco-dlp) [inline hyperspy/component.py]: This `charge` has to come back too.

Comment (francisco-dlp) [inline hyperspy/component.py]: No it doesn't, sorry.

**Pull request merged**